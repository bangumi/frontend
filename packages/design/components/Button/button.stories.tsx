import type { Meta, StoryFn } from '@storybook/react';
import type { FC } from 'react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';

import { ArrowRightCircle } from '@bangumi/icons';

import type { ButtonLinkProps, ButtonProps } from '.';
import Button from '.';

const storyMeta: Meta<typeof Button> = {
  title: 'modern/Button',
  component: Button,
  subcomponents: { 'Button.Link': Button.Link as FC<unknown> },
  argTypes: {
    disabled: { control: 'boolean' },
    color: { control: 'select', options: ['default', 'blue', 'gray'] },
  },
  parameters: {
    docs: {
      description: {
        component:
          '主类型按钮，全圆角，有颜色填充，为默认类型。默认为粉色，另有蓝色和灰色两种颜色可选。',
      },
    },
  },
};

export default storyMeta;

const Template: StoryFn<ButtonProps> = (args) => {
  return <Button {...args}>{args.children ?? 'Click Me!'}</Button>;
};

export const Primary = Template.bind({});
Primary.args = {
  type: 'primary',
  children: '登录',
};

export const Secondary = Template.bind({});
Secondary.args = {
  type: 'secondary',
  children: '小组概览',
  size: 'medium',
};
Secondary.parameters = {
  docs: {
    description: {
      story:
        '次级按钮，全圆角，有边框，无颜色填充。默认为灰色，另有蓝色变种可选，此处为 `medium` 大小。',
    },
  },
};

export const Text = Template.bind({});
Text.args = {
  type: 'text',
  children: '编辑',
  size: 'small',
};
Text.parameters = {
  docs: {
    description: {
      story: '纯文字按钮，无边框，无颜色填充。默认为灰色，无其他颜色变种，此处为 `small` 大小。',
    },
  },
};

export const Blue = Template.bind({});
Blue.args = {
  type: 'primary',
  children: '发表新主题',
  color: 'blue',
};
Blue.parameters = {
  docs: {
    description: {
      story: '主类型按钮的蓝色变种，为默认的 `large` 大小。',
    },
  },
};

export const Small = Template.bind({});
Small.args = {
  type: 'secondary',
  children: '评论',
  size: 'small',
};
Small.parameters = {
  docs: {
    description: {
      story: '所有类型的按钮均提供 `large`、`medium`、`small` 三种尺寸。默认为 `large`。',
    },
  },
};

export const Square = Template.bind({});
Square.args = {
  shape: 'square',
};

export const Disabled = Template.bind({});
Disabled.args = {
  type: 'primary',
  disabled: true,
  children: 'Disabled',
};

const ButtonLinkTemplate: StoryFn<ButtonLinkProps> = (args) => {
  return (
    <MemoryRouter>
      <Button.Link {...args}>{args.children}</Button.Link>
    </MemoryRouter>
  );
};

export const Link = ButtonLinkTemplate.bind({});
Link.args = {
  to: 'https://bgm.tv',
  isExternal: true,
  target: '_blank',
  children: 'Bangumi 番组计划',
};
Link.parameters = {
  docs: {
    description: {
      story: '按钮的链接版本，使用方法与 `Typography.Link` 相同。',
    },
  },
};

export const Plain = ButtonLinkTemplate.bind({});
Plain.args = {
  to: 'https://bgm.tv',
  isExternal: true,
  target: '_blank',
  type: 'plain',
  children: (
    <>
      更多小组成员
      <ArrowRightCircle />
    </>
  ),
};
Plain.parameters = {
  docs: {
    description: {
      story: '无内边距、边框及圆角，最小高度。',
    },
  },
};
