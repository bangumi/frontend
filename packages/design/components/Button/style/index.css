/* 
    Better to use SASS / LESS
*/
.bgm-button {
  width: 100px;
  height: 50px;
  border: 0;
  border-radius: 10px;
  background-color: #fd979b;
  color: white;
  cursor: pointer;
}

.bgm-button__disabled {
  cursor: default;
}
