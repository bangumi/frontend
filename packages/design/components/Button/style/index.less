@import '../../../theme/base';

.bgm-button {
  cursor: pointer;
  user-select: none;
  font-size: 14px;
  font-weight: 600;
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  display: flex;
  gap: 0.25rem;
  align-items: center;
  text-align: center;
  justify-content: center;
  box-sizing: border-box;
  border: 2px transparent solid;
  height: var(--height);
  line-height: var(--height);
  border-radius: calc(var(--height) / 2);

  // 默认大小：large
  --height: 38px;

  padding: 0 32px;

  &--primary {
    background-color: @primary-color;
    color: #fff;

    &.bgm-button--color {
      &-blue {
        background-color: @blue-100;
      }

      &-gray {
        background-color: @gray-10;
        color: @gray-80;
      }
    }
  }

  &--secondary {
    background-color: transparent;
    border-color: @gray-10;
    color: @gray-60;

    &:hover {
      background-color: @gray-10;
      color: @gray-80;
    }

    &.bgm-button--color {
      &-blue {
        border-color: @blue-100;
        color: @blue-100;

        &:hover {
          background-color: @blue-100;
          color: #fff;
        }
      }
    }
  }

  &--text {
    .text-button-mixin();
  }

  &--size {
    &-medium {
      --height: 30px;

      padding: 0 13px;
    }

    &-small {
      --height: 24px;

      padding: 0 17px;
    }
  }

  &--shape-square {
    border-radius: 0;
  }

  &--disabled {
    cursor: default;
  }

  &--link {
    display: inline-flex;
    text-decoration: none;
  }

  &--plain {
    padding: 0;
    border-radius: 0;
    border: none;
    height: auto;
    line-height: 14px;

    .text-button-mixin();
  }
}

.text-button-mixin() {
  background-color: transparent;
  color: @gray-60;

  &:hover {
    color: @blue-100;
  }
}
