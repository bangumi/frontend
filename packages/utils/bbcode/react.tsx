import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
