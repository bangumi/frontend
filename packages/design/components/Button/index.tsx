import './style';

import classNames from 'classnames';
import type { MouseEventHandler } from 'react';
import React from 'react';

import type { LinkProps } from '../Typography';
import Typography from '../Typography';

export interface ButtonCommonProps {
  type?: 'primary' | 'secondary' | 'text' | 'plain';
  shape?: 'square' | 'rounded';
  size?: 'large' | 'medium' | 'small';
  color?: 'default' | 'blue' | 'gray';
  /** 无内边距、边框及圆角，最小高度 */
}

export type ButtonProps = Omit<JSX.IntrinsicElements['button'], 'type' | 'onClick'> &
  ButtonCommonProps & {
    onClick?: MouseEventHandler; // desserts for story book
    htmlType?: JSX.IntrinsicElements['button']['type'];
  };

const Button = ({
  disabled = false,
  className,
  type = 'primary',
  shape = 'rounded',
  size = 'large',
  color = 'default',
  children,
  htmlType,
  ...rest
}: ButtonProps) => {
  return (
    <button
      className={classNames(
        'bgm-button',
        className,
        { 'bgm-button--disabled': disabled },
        `bgm-button--${type}`,
        {
          [`bgm-button--shape-${shape}`]: shape !== 'rounded',
          [`bgm-button--size-${size}`]: size !== 'large',
          [`bgm-button--color-${color}`]: color !== 'default',
        },
      )}
      type={htmlType}
      disabled={disabled}
      {...rest}
    >
      {children}
    </button>
  );
};

export type ButtonLinkProps = LinkProps & ButtonCommonProps;

export const ButtonLink = ({
  children,
  className,
  type = 'primary',
  shape = 'rounded',
  size = 'large',
  color = 'default',
  ...props
}: ButtonLinkProps) => {
  return (
    <Typography.Link
      className={classNames('bgm-button', className, 'bgm-button--link', `bgm-button--${type}`, {
        [`bgm-button--shape-${shape}`]: shape !== 'rounded',
        [`bgm-button--size-${size}`]: size !== 'large',
        [`bgm-button--color-${color}`]: color !== 'default',
      })}
      noStyle
      {...props}
    >
      {children}
    </Typography.Link>
  );
};

Button.Link = ButtonLink;

export default Button;
