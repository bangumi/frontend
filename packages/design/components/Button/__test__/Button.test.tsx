import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import Button from '../index';

describe('Primary Button', () => {
  it.each(['large', 'medium', 'small'] as const)('should render button of size %s', (size) => {
    const { container } = render(
      <Button type='primary' size={size}>
        hello world
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it.each(['blue', 'gray'] as const)('should render button of color %s', (color) => {
    const { container } = render(
      <Button type='primary' color={color}>
        hello world
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('Secondary Button', () => {
  it.each(['large', 'medium', 'small'] as const)('should render button of size %s', (size) => {
    const { container } = render(
      <Button type='secondary' size={size}>
        hello world
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it.each(['blue'] as const)('should render button of color %s', (color) => {
    const { container } = render(
      <Button type='secondary' color={color}>
        hello world
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

describe('Text Button', () => {
  it.each(['large', 'medium', 'small'] as const)('should render button of size %s', (size) => {
    const { container } = render(
      <Button type='text' size={size}>
        hello world
      </Button>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});

it('should render button link', () => {
  const { container } = render(
    <Button.Link to='https://bgm.tv' isExternal>
      Bangumi
    </Button.Link>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('should render plain button', () => {
  const { container } = render(
    <Button.Link to='https://bgm.tv' type='plain' isExternal>
      Bangumi
    </Button.Link>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it.each(['square', 'rounded'] as const)('should render button of shape %s', (shape) => {
  const { container } = render(
    <Button type='primary' shape={shape}>
      hello world
    </Button>,
  );
  expect(container.firstChild).toMatchSnapshot();
});

it('should trigger onClick callback', () => {
  const onClick = vi.fn();
  const { getByText } = render(
    <Button type='primary' onClick={onClick}>
      hello world
    </Button>,
  );
  fireEvent.click(getByText('hello world'));

  expect(onClick).toBeCalledTimes(1);
});

it('should disable the button if disabled is true', () => {
  const { getByText } = render(
    <Button type='primary' disabled>
      hello world
    </Button>,
  );
  expect(getByText('hello world')).toBeDisabled();
});
