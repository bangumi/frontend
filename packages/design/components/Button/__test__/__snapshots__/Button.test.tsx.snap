// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Primary Button > should render button of color blue 1`] = `
<button
  class="bgm-button bgm-button--primary bgm-button--color-blue"
>
  hello world
</button>
`;

exports[`Primary Button > should render button of color gray 1`] = `
<button
  class="bgm-button bgm-button--primary bgm-button--color-gray"
>
  hello world
</button>
`;

exports[`Primary Button > should render button of size large 1`] = `
<button
  class="bgm-button bgm-button--primary"
>
  hello world
</button>
`;

exports[`Primary Button > should render button of size medium 1`] = `
<button
  class="bgm-button bgm-button--primary bgm-button--size-medium"
>
  hello world
</button>
`;

exports[`Primary Button > should render button of size small 1`] = `
<button
  class="bgm-button bgm-button--primary bgm-button--size-small"
>
  hello world
</button>
`;

exports[`Secondary Button > should render button of color blue 1`] = `
<button
  class="bgm-button bgm-button--secondary bgm-button--color-blue"
>
  hello world
</button>
`;

exports[`Secondary Button > should render button of size large 1`] = `
<button
  class="bgm-button bgm-button--secondary"
>
  hello world
</button>
`;

exports[`Secondary Button > should render button of size medium 1`] = `
<button
  class="bgm-button bgm-button--secondary bgm-button--size-medium"
>
  hello world
</button>
`;

exports[`Secondary Button > should render button of size small 1`] = `
<button
  class="bgm-button bgm-button--secondary bgm-button--size-small"
>
  hello world
</button>
`;

exports[`Text Button > should render button of size large 1`] = `
<button
  class="bgm-button bgm-button--text"
>
  hello world
</button>
`;

exports[`Text Button > should render button of size medium 1`] = `
<button
  class="bgm-button bgm-button--text bgm-button--size-medium"
>
  hello world
</button>
`;

exports[`Text Button > should render button of size small 1`] = `
<button
  class="bgm-button bgm-button--text bgm-button--size-small"
>
  hello world
</button>
`;

exports[`should render button link 1`] = `
<a
  class="bgm-button bgm-button--link bgm-button--primary"
  href="https://bgm.tv"
>
  Bangumi
</a>
`;

exports[`should render button of shape rounded 1`] = `
<button
  class="bgm-button bgm-button--primary"
>
  hello world
</button>
`;

exports[`should render button of shape square 1`] = `
<button
  class="bgm-button bgm-button--primary bgm-button--shape-square"
>
  hello world
</button>
`;

exports[`should render plain button 1`] = `
<a
  class="bgm-button bgm-button--link bgm-button--plain"
  href="https://bgm.tv"
>
  Bangumi
</a>
`;
